docs/code/IDataViewDesignPrinciples.md

```diff
@@ -0,0 +1,479 @@
+  When this returns `null`, the row count is not available directly from the
+  view.
+
+* **Efficient indexed row access**: There is no standard way in the IDataView
+  system to request the values for a specific row number. While the
+  `IRowCursor` interface has a `MoveMany(long count)` method, it only supports
+  moving forward `(count > 0)`, and is not necessarily more efficient than
+  calling `MoveNext()` repeatedly. See [here](#row-cursor).
+
+* **Data file formats**: The IDataView system does not dictate storage or
+  transport formats. It *does* include interfaces for loader and saver
+  components. The AzureML Algorithms team has implemented loaders and savers
+  for some binary and text file formats, but additional loaders and savers can
+  (and will) be implemented. In particular, implementing a loader from XDF
+  will be straightforward. Implementing a saver to XDF will likely require the
+  XDF format to be extended to support vector-valued columns.
+
+* **Multi-node computation over multiple data partitions**: The IDataView
+  design is focused on single node computation. We expect that in multi-node
+  applications, each node will be given its own data partition(s) to operate
+  on, with aggregation happening outside an IDataView pipeline.
+
+## Schema and Type System
+
+### Basics
+
+IDataView has general schema support, in that a view can have an arbitrary
+number of columns, each having an associated name, index, data type, and
+optional metadata.
+
+Column names are case sensitive. Multiple columns can share the same name, in
+which case, one of the columns hides the others, in the sense that the name
+will map to one of the column indices, the visible one. All user interaction
+with columns should be via name, not index, so the hidden columns are
+generally invisible to the user. However, hidden columns are often useful for
+diagnostic purposes.
+
+The set of supported column data types forms an open type system, in the sense
+that additional types can be added at any time and in any assembly. However,
+there is a precisely defined set of standard types including:
+
+* Text
+* Boolean
+* Single and Double precision floating point
+* Signed integer values using 1, 2, 4, or 8 bytes
+* Unsigned integer values using 1, 2, 4, or 8 bytes
+* Unsigned 16 byte values for ids and probabilistically unique hashes
+* Date time, date time zone, and timespan
+* Key types
+* Vector types
+
+The set of standard types will likely be expanded over time.
+
+The IDataView type system is specified in a separate document, *IDataView Type
+System Specification*.
+
+IDataView provides a general mechanism for associating semantic metadata with
+columns, such as designating sets of score columns, names associated with the
+individual slots of a vector-valued column, values associated with a key type
+column, whether a column's data is normalized, etc.
+
+While IDataView schema supports an arbitrary number of columns, it, like most
```

Comment:
What's the limits of the "arbitrary" number of columns? int.MaxValue?

Reply:
I think it's arbitrary in the sense that the schema is not fixed, not arbitrary in the sense of completely unlimited and unbounded.

---
In reply to: [188898484](https://github.com/dotnet/machinelearning/pull/173#discussion_r188898484) [](ancestors = 188898484)